### DISCUSSION TITLE
[1주차] 애플리케이션 계층 2.1~4

### DISCUSSION THREAD

**Discussion by H-sooyeon** [Dijkstra-Employ/Network-Study#4] [Q&A]
1. Q: Web Cache (Proxy Server)의 종류와 proxy를 사용할 때의 장점과 단점은 무엇인가요?

2. Q: SMTP가 푸시(push) 프로토콜인 반면, 메시지를 얻는 것은 풀(pull) 동작이라는 것은 어떤 것을 의미하나요? 
         SMTP 푸시와 풀 동작에 대해 설명해주세요.

3. Q:  HTTP 멱등성이란 무엇인가요?

**Comment by Starlight258** [comment]
[1주차]
# Q1.
소켓에 대해 간단하게 설명해보세요.

# Q2. 
네트워크는 `TCP`와 `UDP`가 공존합니다. 
`TCP`는 네트워크가 혼잡 상태일때 프로세스 속도를 낮추는 혼잡 제어를 합니다. 
혼잡 제어를 하지 않는 `UDP`에 비해 `TCP`가 불리하지 않을까요? 
네트워크는 어떻게 TCP와 UDP가 공평하게 작동될 수 있도록 동작할까요? 

# Q3.
캐싱된 데이터를 요청할때 브라우저는 `조건부 GET` 요청을 보냅니다.
브라우저는 어떻게 클라이언트가 요청한 데이터가 캐싱되었는지 캐시 서버에 접근하지 않고 알 수 있을까요?

**Reply by H-sooyeon** [reply]
> Q1 소켓에 대해 간단하게 설명해보세요.

네트워크에서 응용 계층 프로세스들이 소켓을 거쳐 전송 계층으로 데이터를 전달합니다.

즉, 소켓은 전송 계층과 응용 프로그램 사이의 인터페이스 역할을 하며 소켓을 통해 통신이 이루어집니다.

네트워크에서 데이터를 보낸다는 것은 클라이언트 소켓이 서버 소켓에 데이터를 보내는 것을 의미합니다.

<br />

> Q2 네트워크는 어떻게 TCP와 UDP가 공평하게 작동될 수 있도록 동작할까요?

네트워크에서 UDP 송신자들이 TCP 트래픽을 밀어낼 가능성이 생기므로 TCP 관점에서는 UDP상에서 수행되는 애플리케이션은 공평하지 못합니다.

그렇기에 **네트워크 품질(Qos) 설정**, **페어 큐잉(Fair Queuing)**, **트래픽 쉐이핑(Traffic Shaping)** 등의 메커니즘을 사용하여 공평성을 보장하려고 노력합니다.

<br />

**네트워크 품질 Quality of Service**
- 네트워크 트래픽의 우선순위를 지정하는 방법으로, 중요하거나 지연에 민감한 트래픽에 대해 우선순위를 부여하여 다른 트래픽보다 먼저 처리되도록 한다.

**페어 큐잉 Fair Queueing**
- 네트워크 라우터나 스위치가 트래픽을 처리할 때 각 송신자에게 공평하게 대역폭을 분배하여 한 종류의 트래픽이 다른 트래픽을 억누르지 않도록 한다.

**트래픽 쉐이핑 Traffic Shaping**
- 특정 트래픽의 전송 속도를 제한하여 네트워크의 다른 부분에 과도한 부하가 걸리지 않도록 한다. 이는 특정 종류의 트래픽이 전체 대역폭을 독점하지 않도록 제어할 수 있게 한다.

<br />

> Q3 브라우저는 어떻게 클라이언트가 요청한 데이터가 캐싱되었는지 캐시 서버에 접근하지 않고 알 수 있을까요?

리소스의 마지막 갱신 시각으로 캐시 유효성을 검증하는 `Last-Modified`와 `If-Modified-Since` , 리소스의 식별자를 기준으로 캐시 유효성을 검증하는 `ETag`와 `If-None-Match` 헤더를 요청/응답 메시지에 추가하여 실제 원본 데이터가 수정되었을 때만 리소스를 내려 받도록 합니다.

서버에게 해당 헤더를 포함하여 요청 메시지를 전달하면, 서버는 데이터가 변경되지 않았을 경우 텍스트 메시지만 응답하여 클라이언트가 캐시에서 데이터를 추출하여 사용하도록 합니다.

데이터가 변경되었을 경우엔 서버는 해당 데이터를 포함하여 응답 메시지를 클라이언트에게 전달합니다.

**Comment by choiyoung69** [comment]
**Q1.** 
---
Client-server 아키텍처와 P2P 아키텍처의 주요 차이점을 설명해주세요.

**Q2.**  
---
일반적으로 HTTP는 persistent 연결을 사용합니다. 비지속적 연결을 사용하지 않는 이유를 대표적으로 2가지 설명해주세요.

**Q3**. 
---
DNS 서버는 여러 종류로 나뉩니다. 루트 DNS 서버, TLD DNS 서버, 권한 있는 DNS 서버, 로컬 DNS 서버의 역할을 각각 설명하고, 이들이 어떻게 상호작용하는지 설명해주세요.

**Reply by H-sooyeon** [reply]
> Q1 Client-Server 아키텍처와 P2P 아키텍처의 주요 차이점을 설명해주세요.

Client-Server 구조는 서버가 클라이언트 호스트로부터 객체를 요청받으면 서버는 요청된 객체를 클라이언트 호스트로 보내면서 응답합니다.

클라이언트는 서로 직접적으로 통신하지 않으며 서버를 통해 통신하고, 그렇기에 서버의 주소는 고정되어야 클라이언트가 해당 주소로 찾아갈 수 있습니다.

즉, Client-Server 구조는 서버에 의존하게 됩니다.

하지만 P2P 구조는 서버에 최소로 의존하며 피어(Peer)라는 간헐적으로 연결된 호스트 쌍이 서로 직접 통신하게 합니다.

각 피어들은 자신이 가진 파일을 다른 피어들에게 분배함으로써 클라이언트와 서버의 역할을 모두 수행할 수 있습니다.

<br />

> Q2 HTTP가 비지속 연결을 사용하지 않는 이유를 2가지 설명해주세요.

비지속 연결은 서버가 클라이언트를 기억할 수 없기에 동일한 클라이언트에서 연속적으로 요청이 오면 연결과 연결 해제 과정을 반복하게 되어 자원을 낭비하게 됩니다.

클라이언트는 매번 연결을 새로 하게 되어 발생하는 네트워크 지연으로 인해 좋지 않은 경험을 하게 됩니다.

<br />

> Q3 DNS 계층 구조에 포함되는 DN 서버들의 역할을 설명하고 어떻게 상호작용하는지 설명해주세요.

DNS는 분산 데이터베이스 기반이며 Root NS -> TLD NS -> Authotitative NS -> Local NS 순으로 질의를 날려 클라이언트가 요청한 호스트에 대한 IP 주소를 얻을 수 있습니다.

루트 DNS 서버는 루트 도메인인 `.`을 관장하는 네임 서버로, 질의에 대해 TLD 서버의 IP 주소를 반환합니다.

TLD DNS 서버는 하위 도메인 네임을 관리하며 책임 서버 주소를 반환합니다.

책임 DNS 서버는 자신이 관리하는 도메인 영역의 질의에 대해서는 다른 네임 서버에게 떠넘기지 않고 곧바로 답하며, 로컬 네임 서버가 마지막으로 질의하는 네임 서버입니다.

로컬 DNS 서버는 클라이언트와 맞닿아 있는 네임 서버로, 가장 먼저 찾게 되는 네임 서버입니다. 로컬 네임 서버에 찾고자 하는 IP 주소가 없다면, DNS 질의를 시작합니다.

**Comment by choiyoung69** [comment]
Q: Web Cache (Proxy Server)의 종류와 proxy를 사용할 때의 장점과 단점은 무엇인가요?
---
→ Web Cache에는 브라우저 캐시, 프록시 웹 캐시, CDN, ISPs 등이 있다. Web Cache는 클라이언트 요청에 대한 응답시간을 크게 줄일 수 있다는 점과 기관의 인터넷 액세스 링크 트래픽을 상당히 줄일 수 있어서 비용이 절감된다는 장점이 있다. 또한, 인터넷 전체의 웹 트래픽을 크게 줄여 애플리케이션의 성능을 향상시킨다는 장점이 있다. 그러나, 캐시의 복사본이 오래될 수 있다는 단점이 존재한다. + 또한, 프록시 서버를 거치면서 추가적인 응답 시간이 발생할 수 있다는 단점도 있다.

Q: SMTP가 푸시(push) 프로토콜인 반면, 메시지를 얻는 것은 풀(pull) 동작이라는 것은 어떤 것을 의미하나요? SMTP 푸시와 풀 동작에 대해 설명해주세요.
---
SMTP는 push 프로토콜로 송신자의 메일 서버에서 수신지의 메일서버로 정보를 밀어넣는다(push). 그러나 메세지를 얻을 때는 SMTP를 사용하지 않고 HTTP나 IMAP을 사용하는데 이는 client가 요청으로 주기적으로 서버에 연결하여 최근 이벤트를 확인하고 가져 오는 pull 동작이다.

Q: HTTP 멱등성이란 무엇인가요?
---
HTTP 멱등성이란 여러 번 동일한 요청을 보냈을 때, 서버에 미치는 의도된 영향이 동일한 경우를 의미한다. 멱등성을 가진 메서드는 GET, HEAD, DELETE, OPTIONS 등이 있고, 멱등성을 가지지 않는 메서드에는 POST와 PATCH 등이 있다. 만약 HTTP의 요청이 멱등하다면 요청이 실패한 경우 재시도 요청을 하면 되지만 멱등하지 않다면 문제가 발생할 수 있기 때문에 멱등성을 고려해야 한다.

Q: 소켓에 대해 간단하게 설명해보세요.
---
소켓은 호스트 내의 애플리케이션과 네트워크 사이의 API로, 네트워크 프로토콜을 통해 데이터를 송수신하는데 사용된다. 프로세스들은 데이터 송수신을 위해서 소켓을 사용해 메세지를 보내고 받는다.

Q: 네트워크는 `TCP`와 `UDP`가 공존합니다. `TCP`는 네트워크가 혼잡 상태일때 프로세스 속도를 낮추는 혼잡 제어를 합니다. 혼잡 제어를 하지 않는 `UDP`에 비해 `TCP`가 불리하지 않을까요? 네트워크는 어떻게 TCP와 UDP가 공평하게 작동될 수 있도록 동작할까요?
---
네트워크에서는 TCP와 UDP가 공존하기 때문에 네트워크 혼잡 상황에서 TCP가 UDP에 비해 불리한 것처럼 보일 수 있다. 그러나 네트워크는 TCP와 UDP가 공정하게 동작할 수 있도록 여러 메카니즘을 사용한다. 예를 들어, AQM 알고리즘을 통해서 라우터에서 큐를 관리하여 UDP가 TCP 트래픽을 압도하는 것을 방지할 수 있다. 또한 DiffServ는 각 클래스에 우선순위를 부여하는 일을 하는데 TCP 기반트래픽은 UDP에 비해 우선순위를 높게 측정하면서 네트워크 자원을 공평하게 사용할 수 있도록 도와준다. 이외에도 Fair Queueing 등 여러 장치를 통해 공정한 네트워크 상황을 지원한다.

Q. 캐싱된 데이터를 요청할때 브라우저는 `조건부 GET` 요청을 보냅니다. 브라우저는 어떻게 클라이언트가 요청한 데이터가 캐싱되었는지 캐시 서버에 접근하지 않고 알 수 있을까요?
---
캐시는 개체를 요청하는 브라우저로 전달하지만 개체를 로컬로 캐시하기 때문에 브라우저는 로컬 캐시를 확인하여 요청한 데이터가 캐싱되었는지 확인할 수 있다.

Q. Client-server 아키텍처와 P2P 아키텍처의 주요 차이점을 설명해주세요.
---
client-server 아키텍처는 서비스를 요청하는 클라이언트와 서비스를 제공하는 서버로 구성된다. 보통 데이터 센터를 통해 여러 서버를 운영한다. P2P 아키텍처는 네트워크에 연결된 모든 장치가 동등하게 직접 통신하며 전용 서버를 거치지 않는다. 

Q. 일반적으로 HTTP는 persistent 연결을 사용합니다. 비지속적 연결을 사용하지 않는 이유를 대표적으로 2가지 설명해주세요.
---
비지속적 연결은 요청된 각 객체에 대해 새로운 연결을 설정하고 유지해야 하기 때문에 웹 서버에 상당한 부담을 줄 수 있다. 또한, 각 개체는 두 개의 RTT의 전달 지연을 추가로 초래할 수 있다.

Q. DNS 서버는 여러 종류로 나뉩니다. 루트 DNS 서버, TLD DNS 서버, 권한 있는 DNS 서버, 로컬 DNS 서버의 역할을 각각 설명하고, 이들이 어떻게 상호작용하는지 설명해주세요.
---
루트 DNS 서버는 DNS 계층 구조의 최상위에 위치하며, 최상위 도메인(TLD) 서버의 위치를 안내한다. TLD DNS 서버는 특정 최상위 도메인과 관련된 정보를 관리한다.(.com, .org, .kr 등) 주어진 TLD에 대한 권한 있는 DNS 서버의 주소를 제공한다. 권한 있는 DNS 서버는 특정 도메인에 대한 권한 있는 정보를 제공한다. 즉, 도메인 이름과 관련된 최종적인 IP 주소를 제공한다. 로컬 DNS 서버는 일반적으로 ISP나 기업 네트워크에 의해 운영되며 캐시를 사용하여 최근에 조회된 도메인 이름과 IP 주소를 저장하고 응답 속도를 향상시킨다.

→ 클라이언트가 도메인 이름을 입력하면 로컬 DNS 서버는 캐시에서 해당 IP 주소를 찾는다. 만약 IP 주소가 없다면 DNS 요청을 처리하기 위해 루트 DNS 서버부터 시작하여 TLD DNS 서버, 권한 있는 DNS 서버를 차례로 조회한다. 이후, 권한 있는 DNS 서버로부터 IP 주소를 받으면 로컬 DNS 서버는 이 정보를 캐시에 저장하고 클라이언트에게 응답한다.

**Comment by Starlight258** [comment]
> 1. Q: Web Cache (Proxy Server)의 종류와 proxy를 사용할 때의 장점과 단점은 무엇인가요?
- Forward Proxy : 클라이언트 측에 위치하여 클라이언트의 요청을 대신 받아 서버에 전달
- Reverse Proxy: 서버 측에 위치하여 클라이언트의 요청을 받아 적절한 백엔드 서버로 전달
- 추가로 CDN, Transparent Proxy도 존재

- 프록시를 사용하면 자주 요청되는 콘텐츠를 캐시하여 응답 시간이 단축되며, 클라이언트의 ip를 숨기거나 필터링, 접근 제어, 부하 분산등을 수행할 수 있다.
> 2. Q: SMTP가 푸시(push) 프로토콜인 반면, 메시지를 얻는 것은 풀(pull) 동작이라는 것은 어떤 것을 의미하나요?
>    SMTP 푸시와 풀 동작에 대해 설명해주세요.
- SMTP는 파일을 송신할 컴퓨터가 먼저 연결을 초기화하므로 push(메세지 push) 프로토콜이다. 
  수신자는 메세지를 얻기 위해 사용자 에이전트를 통해 메일 서버의 메세지를 가져오는 pull 동작을 수행한다.
  메세지를 가져오기 위해 메일 접속 프로토콜(pull 프로토콜)인 POP3나 IMAP을 사용한다.
  
> 3. Q:  HTTP 멱등성이란 무엇인가요?
- 동일한 요청을 여러번 수행해도 서버의 상태가 한번 수행한 것과 같은 상태가 되는 성질이며 GET, HEAD, PUT, DELETE, OPTIONS가 이에 해당한다.

> ## **Q1.**
> Client-server 아키텍처와 P2P 아키텍처의 주요 차이점을 설명해주세요.
 - 클라이언트-서버 구조에서 클라이언트는 항상 요청을 보내는 쪽이고, 서버는 항상 서비스를 제공하는 쪽이다. (중앙 집중형)
 - P2P 구조에서 모든 노드는 서버이자 클라이언트이며 동등한 권한과 책임을 가진다. (분산형)

> ## **Q2.**
> 일반적으로 HTTP는 persistent 연결을 사용합니다. 비지속적 연결을 사용하지 않는 이유를 대표적으로 2가지 설명해주세요.
> 
1. 성능 향상
- 비지속 연결을 사용할 경우 모든 요구-응답 쌍이 분리된 TCP 연결을 통해 보내져야하므로 비효율적이다.
- 지속 연결을 사용하면 한번 설정한 TCP 연결을 여러 요청에 재사용 가능하므로 매번 핸드쉐이킹을 하지 않아도 되어 효율적이다. 
2. 서버 리소스 소모 줄임
- 비지속 연결을 사용할 경우 각 연결마다 매번 서버에서 리소스를 할당하고 해제해야하지만, 지속적 연결을 사용하면 서버의 소켓 및 메모리 사용을 줄일 수 있다.
- > HTTP/1.1 부터 기본적으로 지속 연결 사용

> ## **Q3**.
> DNS 서버는 여러 종류로 나뉩니다. 루트 DNS 서버, TLD DNS 서버, 권한 있는 DNS 서버, 로컬 DNS 서버의 역할을 각각 설명하고, 이들이 어떻게 상호작용하는지 설명해주세요.

- 루트 DNS 서버: 계층 구조의 최상위에 위치하며 TLD DNS 서버의 IP 주소 제공
- TLD DNS 서버: 상위 레벨 도메인과 국가의 상위 레벨 도메인에 대한 TLD 서버가 있다. 해당 TLD에 속한 책임 DNS 서버의 IP 주소가 있다.
- 책임 DNS 서버: 특정 도메인에 대한 실제 DNS 레코드를 보유하며 호스트 네임을 IP 주소로 변환하는 정보를 제공한다.

- 상호 작용
호스트가 로컬 DNS 서버에 DNS 질의를 보내고, 루트 DNS 서버가 TLD 서버를 캐싱할 경우 TLD 서버에, 그렇지 않을 경우 루트 DNS 서버에 질의를 보낸 후 응답받아 TLD DNS 서버에 질의를 보낸다. TLD DNS 서버는 책임 DNS 서버의 IP 주소를 보내고 책임 DNS 서버에 질의하여 헤당 도메인에 대한 IP 주소를 얻는다.
> 요청하는 호스트~로컬 DNS 서버까지의 질의는 재귀적이고, 나머지는 반복적 질의이다.

> # Q1.
> 소켓에 대해 간단하게 설명해보세요.
- 소켓이란 호스트의 애플리케이션 프로세스와 전송 계층 프로토콜 사이의 인터페이스이다. 
송신 측의 애플리케이션은 소켓을 통해 메세지를 보내며, 전송 프로토콜은 그 메세지를 수신 프로세스의 소켓으로 이동시키는 책임을 가진다.

> # Q2.
> 네트워크는 `TCP`와 `UDP`가 공존합니다. `TCP`는 네트워크가 혼잡 상태일때 프로세스 속도를 낮추는 혼잡 제어를 합니다. 혼잡 제어를 하지 않는 `UDP`에 비해 `TCP`가 불리하지 않을까요? 네트워크는 어떻게 TCP와 UDP가 공평하게 작동될 수 있도록 동작할까요?
> 
- 상당 부분 라우터에서 수행한다.
  - 라우터에서 큐를 관리한다. (공정 큐잉-각 흐름에 동등한 대역폭 할당, 가중치 기반 공정 큐잉-우선순위에 따른 대역폭 할당)
  - 패킷스케줄링(패킷의 전송 순서 조정)
  - 액티브 큐 관리
-  그 외 네트워크 장비
   - QoS 정책: 특정 유형의 트래픽에 우선순위 부여(TCP와 UDP 트래픽의 균형 조절)
   - 대역폭 제한(UDP 트래픽에 대한 대역폭 상한선 설정)
> # Q3.
> 캐싱된 데이터를 요청할때 브라우저는 `조건부 GET` 요청을 보냅니다. 브라우저는 어떻게 클라이언트가 요청한 데이터가 캐싱되었는지 캐시 서버에 접근하지 않고 알 수 있을까요?
- 브라우저는 자체 로컬 캐시를 사용하며 리소스를 처음 받을때 리소스와 관련 메타 정보를 로컬 캐시에 저장한다.
- 사용자가 리소스를 요청하면 해당 리소스가 로컬 캐시에 있는지 확인하고, 캐시된 리소스가 있다면 캐시의 만료 여부를 확인한다. (max-age or Expires or Last-Modified 기반 휴리스틱 만료 계산)
- 만료되지 않았다면 캐시된 리소스를 사용하고, 만료되었다면 조건부 GET 요청을 서버로 전송한다.
